Optimization for BigInt by introducing BigInt64

hyjorc1 opened the pull request:
#### a7f2c6afa7c7413e2ae4e78b8bbca85010591f22
<pre>
Optimization for BigInt by introducing BigInt64
Need the bug URL (OOPS!).
<a href="https://rdar.apple.com/105473877">rdar://105473877</a>

Reviewed by NOBODY (OOPS!).

The operations for HeapBigInt in DFG and FTL are expensive due to the
heap storage. In order to leverage benefits of arithmetic computation
under the hood, this patch introduces BigInt64 as a representation for
HeapBigInt within 64-bit range. Note that BigInt64 is not a JSValue,
it&apos;s an internal representation used for computation in DFG and FTL.

Given DFG graph:

D@1: GetLocal(SpeculatedType: HeapBigInt)
D@2: GetLocal(SpeculatedType: HeapBigInt)
D@3: ValueAdd(HeapBigIntUse:D@1, HeapBigIntUse:D@2, NodeFlags: MayHaveHeapBigIntResult, SpeculatedType: ?)
D@4: Return(Untyped:D@3)

At DFGPredictionPropagationPhase, check D@3 and update its SpeculatedType to SpecBigInt64.

D@1: GetLocal(SpeculatedType: HeapBigInt)
D@2: GetLocal(SpeculatedType: HeapBigInt)
D@3: ValueAdd(HeapBigIntUse:D@1, HeapBigIntUse:D@2, NodeFlags: MayHaveHeapBigIntResult, SpeculatedType: SpecBigInt64)
D@4: Return(Untyped:D@3)

At DFGFixupPhase, once we saw D@3 has SpecBigInt64 type then:
1. Fix D@3&apos;s edges&apos; uses to BigInt64RepUse. With that D@5 and D@6 can be inserted since we want to introduce ArithAdd to D@3.
2. Update D@3&apos;s node flag to NodeResultBigInt64. Meanwhile, a HeapBigInt ValueRep should be applied for a valid JSValue.

D@1: GetLocal(SpeculatedType: HeapBigInt)
D@2: GetLocal(SpeculatedType: HeapBigInt)
D@5: BigInt64Rep(HeapBigIntUse:D@1, NodeFlags: NodeResultBigInt64, SpeculatedType: SpecBigInt64)
D@6: BigInt64Rep(HeapBigIntUse:D@2, NodeFlags: NodeResultBigInt64, SpeculatedType: SpecBigInt64)
D@3: ArithAdd(BigInt64RepUse:D@5, BigInt64RepUse:D@6, NodeFlags: NodeResultBigInt64, SpeculatedType: SpecBigInt64)
D@7: ValueRep(BigInt64RepUse:D@3, NodeFlags: JS|PureInt, SpeculatedType: HeapBigInt)
D@4: Return(Untyped:D@7)

There is a tradeoff for emitting more BigInt64Rep or ValueRep (of HeapBigInt) nodes.
Based on our observation, ValueRep is far more expensive than BigInt64Rep. So,
the current optimization strategy applied to this patch is to emit ValueRep nodes
as few as possible.

Microbenchmark Results:

                                        without                     with

bigint64-inc-single                  2.4576+-0.0599     ^      1.4783+-0.0566        ^ definitely 1.6624x faster
bigint64-or-single                   3.0919+-0.0339     ^      1.5873+-0.0257        ^ definitely 1.9478x faster
bigint64-mul-multiple               16.8231+-0.7327     ^      1.6956+-0.0267        ^ definitely 9.9215x faster
bigint64-less                        8.4980+-0.1206     ^      1.0925+-0.1443        ^ definitely 7.7783x faster
bigint64-mul-single                  4.8715+-0.0955     ^      1.5573+-0.0388        ^ definitely 3.1281x faster
bigint64-not-multiple                7.4200+-0.2303     ?      7.5734+-0.8582        ? might be 1.0207x slower
bigint64-xor-multiple               11.3447+-0.3092     ^      1.6862+-0.1020        ^ definitely 6.7279x faster
bigint64-sub-single                  3.9847+-0.2184     ^      1.5923+-0.0152        ^ definitely 2.5025x faster
bigint64-mod-multiple                2.5306+-0.0594     ^      1.5640+-0.0812        ^ definitely 1.6180x faster
bigint64-mod-single                  1.6990+-0.0208     ^      1.3683+-0.0571        ^ definitely 1.2417x faster
bigint64-dec-single                  2.4088+-0.0305     ^      1.4402+-0.0247        ^ definitely 1.6725x faster
bigint64-less-eq                     8.5159+-0.3171     ^      1.0651+-0.0057        ^ definitely 7.9957x faster
bigint64-neg-multiple                7.4975+-0.1265     ^      1.5273+-0.0349        ^ definitely 4.9090x faster
bigint64-sub-multiple               14.5333+-0.4067     ^      1.7335+-0.0358        ^ definitely 8.3837x faster
bigint64-dec-multiple                9.8675+-0.1191     ^      1.5552+-0.0151        ^ definitely 6.3447x faster
bigint64-eq                          7.9210+-0.0772     ^      1.0878+-0.0540        ^ definitely 7.2818x faster
bigint64-greater                     8.5974+-0.1489     ^      1.0647+-0.0458        ^ definitely 8.0749x faster
bigint64-right-shift-single          2.7958+-0.0509     ^      1.7130+-0.0422        ^ definitely 1.6321x faster
bigint64-neg-single                  2.7797+-0.0492     ^      1.4150+-0.0315        ^ definitely 1.9644x faster
bigint64-inc-multiple                9.8654+-0.0492     ^      1.5998+-0.1471        ^ definitely 6.1666x faster
bigint64-left-shift-single           2.8746+-0.2389     ^      1.7573+-0.0524        ^ definitely 1.6358x faster
bigint64-pow-single                  4.2000+-0.0837     ^      1.6633+-0.0749        ^ definitely 2.5251x faster
bigint64-add-single                  4.5645+-0.1537     ^      1.6910+-0.0615        ^ definitely 2.6992x faster
bigint64-eq-strict                   7.2746+-0.6813     ^      1.0759+-0.0580        ^ definitely 6.7612x faster
bigint64-or-multiple                 9.8060+-0.0574     ^      1.6422+-0.0076        ^ definitely 5.9712x faster
bigint64-add-multiple               15.2192+-0.6459     ^      1.7153+-0.0088        ^ definitely 8.8728x faster
bigint64-left-shift-multiple         5.5388+-0.2262     ^      1.9615+-0.0351        ^ definitely 2.8237x faster
bigint64-right-shift-multiple        5.1425+-0.1992     ^      2.1516+-0.2384        ^ definitely 2.3901x faster
bigint64-div-multiple                4.1177+-0.1316     ^      1.8295+-0.0835        ^ definitely 2.2508x faster
bigint64-and-single                  3.4117+-0.1789     ^      1.4275+-0.0399        ^ definitely 2.3900x faster
bigint64-and-multiple                8.7764+-0.1988     ^      1.5218+-0.0177        ^ definitely 5.7672x faster
bigint64-pow-multiple                5.3751+-0.0900     ^      1.9319+-0.1145        ^ definitely 2.7822x faster
bigint64-not-single                  2.7936+-0.1594            2.7685+-0.0361
bigint64-div-single                  1.5018+-0.1488     ?      1.5065+-0.0553        ?
bigint64-greater-eq                  8.5748+-0.2135     ^      1.0933+-0.1070        ^ definitely 7.8433x faster
bigint64-xor-single                  3.2157+-0.3507     ^      1.5756+-0.0310        ^ definitely 2.0410x faster

&lt;geometric&gt;                          5.2940+-0.0433     ^      1.6114+-0.0177        ^ definitely 3.2854x faster
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a7f2c6afa7c7413e2ae4e78b8bbca85010591f22

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40764 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19777 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/43142 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/43322 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/36854 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/43071 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22765 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/17108 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/33808 "Found 1 new test failure: media/track/media-element-enqueue-event-crash.html (failure)") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/41338 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16725 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35135 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14413 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/14504 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/36124 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44594 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/34217 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36980 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36437 "Found 1 new test failure: http/tests/navigation/ping-attribute/area-cross-origin-from-https-UpgradeMixedContent.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/40178 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/40390 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/15579 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12783 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/38521 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/17198 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/47400 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/17249 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/9739 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/16842 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a357ad65888ce8e499e96e70ad9e88df0853d67c. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

Constellation commented on Source/JavaScriptCore/bytecode/SpeculatedType.h:
```diff
@@ -102,13 +102,16 @@ static constexpr SpeculatedType SpecMisc                              = SpecBool
 static constexpr SpeculatedType SpecBigInt                            = SpecHeapBigInt;
 #endif
-static constexpr SpeculatedType SpecDataViewObject                    = 1ull << 45; // It's definitely a JSDataView.
+
```
SpecFullTop should include SpecBigInt64

Constellation commented on Source/JavaScriptCore/dfg/DFGOperations.cpp:
```diff
@@ -3279,6 +3279,18 @@ JSC_DEFINE_JIT_OPERATION(operationConvertDoubleToInt52, int64_t, (double value))
+    JITOperationPrologueCallFrameTracer tracer(vm, callFrame);
+
+    JSBigInt* bigInt = jsCast<JSBigInt*>(op);
+    if (bigInt->length() <= 1) // Is this right?
+        return JSBigInt::toBigInt64(bigInt);
```
This is not right since HeapBigInt stores signedness separately.

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.h:
```diff
@@ -797,6 +798,16 @@ class SpeculativeJIT : public JITCompiler {
     {
         int52Result(reg, node, DataFormatStrictInt52, mode);
     }
+    void int64Result(GPRReg reg, Node* node)
```
we should say `bigInt64Result`.

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.h:
```diff
@@ -797,6 +798,16 @@ class SpeculativeJIT : public JITCompiler {
+        GenerationInfo& info = generationInfoFromVirtualRegister(virtualRegister);
+
+        m_gprs.retain(reg, virtualRegister, SpillOrderJS);
+        info.initInt64(node, node->refCount(), reg);
```
Ditto.

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
```diff
@@ -1347,6 +1347,47 @@ GPRReg SpeculativeJIT::fillSpeculateInt52(Edge edge, DataFormat desiredFormat)
+            unlock(gpr);
+            gpr = result;
+        } else
+            info.fillInt64(m_stream, gpr);
```
Ditto.

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
```diff
@@ -3294,6 +3335,33 @@ void SpeculativeJIT::compile(Node* node)
+            GPRReg valueGPR = cell.gpr();
+            speculateHeapBigInt(node->child1(), valueGPR);
+
+            flushRegisters();
+            GPRFlushedCallResult result(this);
+            GPRReg resultGPR = result.gpr();
+            callOperation(operationBigIntToInt64, resultGPR, LinkableConstant::globalObject(*this, node), valueGPR);
+
+            DFG_TYPE_CHECK_WITH_EXIT_KIND(BigInt64Overflow,
+                JSValueRegs(), node->child1(), SpecHeapBigInt,
+                branch64(
+                    Equal, resultGPR,
+                    TrustedImm64(JSBigInt::notInt64)));
```
We should fully inline this code and check overflow correctly.

Constellation commented on Source/JavaScriptCore/runtime/JSBigInt.h:
```diff
@@ -461,6 +461,9 @@ class JSBigInt final : public JSCell {
         return static_cast<int64_t>(toBigUInt64Heap(bigInt.asHeapBigInt()));
     }
 
+    // Use int64 maximum as the flag to indicate to bigint to int64 conversion failure. Note that it's only used for converting JSBigInt to int64_t in DFG/FTL tier for BigInt64.
+    static constexpr const int64_t notInt64 = std::numeric_limits<int64_t>::max();
```
Let's just do inlined check for conversion. And for all Int64 operations, you can check overflow flag in JIT and OSR exit. So we should not have this.

Jarred-Sumner commented:
Assuming most usages of BigInt fit inside 64 bits, I wonder if it'd be worthwhile to have a JSBigInt64 that appears as a bigint to JS but is a JSCell subclass with the value stored as a int64_t. This would save a memory allocation for the bytes, and probably make comparisons & conversions to & from numbers faster. Another idea is something like JSArrayBufferView where it has a union of different internal representations

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/027e0ef441d57fa34a1057eabf5c59be15d47c89. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/027e0ef441d57fa34a1057eabf5c59be15d47c89. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d6d24c524fc0ac94d4930709559ec9dea76a8525. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/18e6502ea878ea2a797c006e11331d45b92bb46e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/18e6502ea878ea2a797c006e11331d45b92bb46e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/1676b385361ed2954788b63d84c9b36cab90d6cd. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/0633c4ca5567aefcd5a767253bd8aa2bc354f86b. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/0633c4ca5567aefcd5a767253bd8aa2bc354f86b. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/aa58e3be521bde41f53dfc2da4cec6b0c86496f5. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/66463e24177f398887daa5ab5d5dbdcdb6377596. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a3ce308d3d4240525e42082e93f4a60309d0eb3a. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d37dd814e819c1679316a5f48e2cee34ed69187d. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/94a15d0cb47aa0e85204b6deb458eb486380543b. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/94a15d0cb47aa0e85204b6deb458eb486380543b. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/0959dead271f92e62853cbaa35ddc2aa3a207379. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/64a8c931624e9dfdf233c2e178c691e01c4d5302. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/47934f87831a493ea4108c58211af68e64d343fe. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/34766209ccb571e97674c462f88d46d609fbba25. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/4664717aa7e592de244a3d4ad64830cf66e94d9c. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/1167e941986aae001e63930fcf4f8f0e6c441013. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/c995b2d4040f65aa2e67196653e5e263ba466039. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/51f23f020d3da6eec5a0463cc714576dd231bdac. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/274eac3523b44d9475ab2253fb3f605e64fca3b0. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/50cf2882b9d21e9ff24dd14744ac81dda7f4147f. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/5102f42485f477c086750f0865e772884fb8f861. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/1d0fa85b3a97adb0a0afbb0016f2ff2a090e4e83. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/7f5d17cd896034fa99a92a43ecad8d36bcd2f976. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/eb7cb36960ae906fe3b67311e9fec0138402dafc. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/001b01ecfbed73477315f8c6efaafebb092b5df4. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/f3eaad010d012150b239ecbd44668267264738ca. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/89fec894f9ba483264335acb0916e5f310e26d91. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/89fec894f9ba483264335acb0916e5f310e26d91. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d8e6dd26c31c7ea6844930102226803846676b77. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/52f34c3d397b1335c944ab18216b5ee3df5155e5. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/43619c1129ad6cfc8f84c47eb530dc4c4311c14e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/8361a69f872fda9f65164b7569234a86ad58ced6. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/6ef0606bdb7198df05bcbabefd0f055e64581a70. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/c715ac52c4fe98616a7be57d035bed9ed5ad5655. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/5e9714c1218717e46aedf610048e7b50a45d5b1d. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/5e9714c1218717e46aedf610048e7b50a45d5b1d. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/c44a320f115980f387d86ed3f5ebc20ad1ba8d6a. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/c44a320f115980f387d86ed3f5ebc20ad1ba8d6a. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/4b5c7d5aff18af98e0ca01a220d4ab0903414768. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/9e615598f1c3bf9d4df8fa329426ff85806b0822. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/f54b598b0c6d32c519a2930110ffde6d6d1ef872. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/83a8171d8c674c08deb7e43c615136e35eabe363. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/83a8171d8c674c08deb7e43c615136e35eabe363. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/04ad70c4ab6f067445080b3df62b206dea0ad040. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/04ad70c4ab6f067445080b3df62b206dea0ad040. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/f3242d9e0d8bfc91290a99f3082d8ac552313e43. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/4c816b82dfd1106c2db473ddaf9d5026171b023a. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/29817e7f8bf6a5df02bbbc8b4bfe094c519f9de6. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/01bbd75c25d80c0cf18de2eb5843d5d33e9657ba. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

hyjorc1 committed:
Optimization for BigInt by introducing BigInt64
Need the bug URL (OOPS!).
rdar://105473877

Reviewed by NOBODY (OOPS!).

The operations for HeapBigInt in DFG and FTL are expensive due to the
heap storage. In order to leverage benefits of arithmetic computation
under the hood, this patch introduces BigInt64 as a representation for
HeapBigInt within 64-bit range. Note that BigInt64 is not a JSValue,
it's an internal representation used for computation in DFG and FTL.

Given DFG graph:

D@1: GetLocal(SpeculatedType: HeapBigInt)
D@2: GetLocal(SpeculatedType: HeapBigInt)
D@3: ValueAdd(HeapBigIntUse:D@1, HeapBigIntUse:D@2, NodeFlags: MayHaveHeapBigIntResult, SpeculatedType: ?)
D@4: Return(Untyped:D@3)

At DFGPredictionPropagationPhase, check D@3 and update its SpeculatedType to SpecBigInt64.

D@1: GetLocal(SpeculatedType: HeapBigInt)
D@2: GetLocal(SpeculatedType: HeapBigInt)
D@3: ValueAdd(HeapBigIntUse:D@1, HeapBigIntUse:D@2, NodeFlags: MayHaveHeapBigIntResult, SpeculatedType: SpecBigInt64)
D@4: Return(Untyped:D@3)

At DFGFixupPhase, once we saw D@3 has SpecBigInt64 type then:
1. Fix D@3's edges' uses to BigInt64RepUse. With that D@5 and D@6 can be inserted since we want to introduce ArithAdd to D@3.
2. Update D@3's node flag to NodeResultBigInt64. Meanwhile, a HeapBigInt ValueRep should be applied for a valid JSValue.

D@1: GetLocal(SpeculatedType: HeapBigInt)
D@2: GetLocal(SpeculatedType: HeapBigInt)
D@5: BigInt64Rep(HeapBigIntUse:D@1, NodeFlags: NodeResultBigInt64, SpeculatedType: SpecBigInt64)
D@6: BigInt64Rep(HeapBigIntUse:D@2, NodeFlags: NodeResultBigInt64, SpeculatedType: SpecBigInt64)
D@3: ArithAdd(BigInt64RepUse:D@5, BigInt64RepUse:D@6, NodeFlags: NodeResultBigInt64, SpeculatedType: SpecBigInt64)
D@7: ValueRep(BigInt64RepUse:D@3, NodeFlags: JS|PureInt, SpeculatedType: HeapBigInt)
D@4: Return(Untyped:D@7)

There is a tradeoff for emitting more BigInt64Rep or ValueRep (of HeapBigInt) nodes.
Based on our observation, ValueRep is far more expensive than BigInt64Rep. So,
the current optimization strategy applied to this patch is to emit ValueRep nodes
as few as possible.

Microbenchmark Results:

                                        without                     with

bigint64-inc-single                  2.4576+-0.0599     ^      1.4783+-0.0566        ^ definitely 1.6624x faster
bigint64-or-single                   3.0919+-0.0339     ^      1.5873+-0.0257        ^ definitely 1.9478x faster
bigint64-mul-multiple               16.8231+-0.7327     ^      1.6956+-0.0267        ^ definitely 9.9215x faster
bigint64-less                        8.4980+-0.1206     ^      1.0925+-0.1443        ^ definitely 7.7783x faster
bigint64-mul-single                  4.8715+-0.0955     ^      1.5573+-0.0388        ^ definitely 3.1281x faster
bigint64-not-multiple                7.4200+-0.2303     ?      7.5734+-0.8582        ? might be 1.0207x slower
bigint64-xor-multiple               11.3447+-0.3092     ^      1.6862+-0.1020        ^ definitely 6.7279x faster
bigint64-sub-single                  3.9847+-0.2184     ^      1.5923+-0.0152        ^ definitely 2.5025x faster
bigint64-mod-multiple                2.5306+-0.0594     ^      1.5640+-0.0812        ^ definitely 1.6180x faster
bigint64-mod-single                  1.6990+-0.0208     ^      1.3683+-0.0571        ^ definitely 1.2417x faster
bigint64-dec-single                  2.4088+-0.0305     ^      1.4402+-0.0247        ^ definitely 1.6725x faster
bigint64-less-eq                     8.5159+-0.3171     ^      1.0651+-0.0057        ^ definitely 7.9957x faster
bigint64-neg-multiple                7.4975+-0.1265     ^      1.5273+-0.0349        ^ definitely 4.9090x faster
bigint64-sub-multiple               14.5333+-0.4067     ^      1.7335+-0.0358        ^ definitely 8.3837x faster
bigint64-dec-multiple                9.8675+-0.1191     ^      1.5552+-0.0151        ^ definitely 6.3447x faster
bigint64-eq                          7.9210+-0.0772     ^      1.0878+-0.0540        ^ definitely 7.2818x faster
bigint64-greater                     8.5974+-0.1489     ^      1.0647+-0.0458        ^ definitely 8.0749x faster
bigint64-right-shift-single          2.7958+-0.0509     ^      1.7130+-0.0422        ^ definitely 1.6321x faster
bigint64-neg-single                  2.7797+-0.0492     ^      1.4150+-0.0315        ^ definitely 1.9644x faster
bigint64-inc-multiple                9.8654+-0.0492     ^      1.5998+-0.1471        ^ definitely 6.1666x faster
bigint64-left-shift-single           2.8746+-0.2389     ^      1.7573+-0.0524        ^ definitely 1.6358x faster
bigint64-pow-single                  4.2000+-0.0837     ^      1.6633+-0.0749        ^ definitely 2.5251x faster
bigint64-add-single                  4.5645+-0.1537     ^      1.6910+-0.0615        ^ definitely 2.6992x faster
bigint64-eq-strict                   7.2746+-0.6813     ^      1.0759+-0.0580        ^ definitely 6.7612x faster
bigint64-or-multiple                 9.8060+-0.0574     ^      1.6422+-0.0076        ^ definitely 5.9712x faster
bigint64-add-multiple               15.2192+-0.6459     ^      1.7153+-0.0088        ^ definitely 8.8728x faster
bigint64-left-shift-multiple         5.5388+-0.2262     ^      1.9615+-0.0351        ^ definitely 2.8237x faster
bigint64-right-shift-multiple        5.1425+-0.1992     ^      2.1516+-0.2384        ^ definitely 2.3901x faster
bigint64-div-multiple                4.1177+-0.1316     ^      1.8295+-0.0835        ^ definitely 2.2508x faster
bigint64-and-single                  3.4117+-0.1789     ^      1.4275+-0.0399        ^ definitely 2.3900x faster
bigint64-and-multiple                8.7764+-0.1988     ^      1.5218+-0.0177        ^ definitely 5.7672x faster
bigint64-pow-multiple                5.3751+-0.0900     ^      1.9319+-0.1145        ^ definitely 2.7822x faster
bigint64-not-single                  2.7936+-0.1594            2.7685+-0.0361
bigint64-div-single                  1.5018+-0.1488     ?      1.5065+-0.0553        ?
bigint64-greater-eq                  8.5748+-0.2135     ^      1.0933+-0.1070        ^ definitely 7.8433x faster
bigint64-xor-single                  3.2157+-0.3507     ^      1.5756+-0.0310        ^ definitely 2.0410x faster

<geometric>                          5.2940+-0.0433     ^      1.6114+-0.0177        ^ definitely 3.2854x faster

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/9f90869d44d9d0903a7f834e570510e81460d415. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a81394058fe40b75170b099dad0353c94293de73. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/c19cb9bbab27884f81a6b5c083e948f53f6df095. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a7f2c6afa7c7413e2ae4e78b8bbca85010591f22. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10658
